rfcs/2023-09-19-18604-metrics-data-model-refactor.md

```diff
@@ -0,0 +1,466 @@
+falsely captured by this heuristic, they would simply be handled as-is at the end of processing.
+
+### Removal of `Set`
+
+We would entirely remove the `Set` variant from `MetricValue`. In order to still be able to handle
+sets coming in, we would handle aggregation at the source itself.
+
+In StatsD, the original model for all metric types was server-local aggregation: the `statsd` daemon
+that you sent metrics to would handle the aggregation, and on a configurable interval, emit the
+aggregated value. This applied to sets as well: the count of unique values in a set, during the
+given aggregation window, was emitted as a gauge. Similarly, the Datadog Agent works this way as
+well, emitting a gauge representing the counter of unique values seen during the aggregation window.
+
+Currently, Vector will handle sets natively, passing along the actual value seen up until it reaches
+a sink, where we either emit the metric directly (in the case of the `statsd` sink) or convert it to
+a gauge, such as in the `datadog_metrics` sink.
+
+With this change, we would move to aggregating set values within the `statsd` source itself and
+emitting them, as a gauge, on an interval. This would accomplish a few things:
+
+- we would remove the need to handle `Set` metrics in all metrics-capable components knowing that,
+  in nearly all cases, they will simply be turned into a gauge before leaving Vector
+- we would emulate StatsD/DogStatsD in terms of aggregation behavior, emitting the gauge directly in
+  the source, which would maintain parity with how StatsD/DogStatsD emit any set metrics
+
+This is not without potential caveats, namely around metric tagging and how those emitted gauges can
+be aggregated downstream (more on that in the Drawbacks section)
+
+### Refactoring `Distribution` into an enum that handled all distribution types
+
+We would refactor the `Distribution` variant to be a dedicated enum that represented all possible
+forms of a "distribution". While there are a few different forms and uses for the word
+"distribution," we're referring to them here as any metric that encapsulates multiple
+samples/observations. This could be the raw samples themselves, or it could be a
+compressed/condensed form such as a histogram or sketch.
+
+One key aspect here is that a distribution to meant to encompass all relevant/possible forms that
+Vector may need to handle and in a way that can be used interchangeably between sources and sinks,
+such as taking in a fixed-bucket histogram from the `prometheus_scrape` source and being able to get
+a DDSketch to send out from the `datadog_metrics` sink.
+
+Said another way, the goal here is to be able to hand a `Distribution` metric to any metrics-capable
+sink and allow it to represent that distribution in the highest fidelity way possible, whether
+that's losslessly converting from one distribution variant to another, or emitting fixed quantiles
```

Comment:
Is it actually possible to losslessly convert between the distribution variants?

Reply:
For the majority of them, yes:

- raw distributions can be converted to either a histogram (fixed or exponential) or a DDSketch
- fixed and exponential histograms can be converted to a DDSketch losslessly although the buckets are still being interpolated, so if the starting buckets are very very wide, you won't pick up any additional accuracy/resolution converting to a DDSketch, but the quantile queries will be statistically accurate

A DDSketch is sort of an exponential histogram under the hood, but because it represents a continuous range vs exponential histograms not doing so, it cannot be converted losslessly.

Finally, nothing can be converted to a raw distribution because original samples have been lost at that point.